onnxruntime/contrib_ops/cuda/bert/fast_gelu_impl.cu

```diff
@@ -0,0 +1,120 @@
+
+__device__ inline float Tanh(const float& x) {
+  return tanhf(x);
+}
+
+__device__ inline half Tanh(const half& x) {
+  const float tmp = tanhf(__half2float(x));
+  return __float2half(tmp);
+}
+
+__device__ inline half2 Tanh(const half2& x) {
+  // at the moment, there is no half2 tanh builtin
+  float2 tmp = (__half22float2(x));
+  tmp.x = tanhf(tmp.x);
+  tmp.y = tanhf(tmp.y);
+  return __float22half2_rn(tmp);
+}
+
+template <typename T, unsigned TPB>
+__global__ void FastGeluKernel(const T a, const T b, const T c, int input_length, int bias_length, const T* input, const T* bias, T* output) {
+  const int idx = blockIdx.x * TPB + threadIdx.x;
+
+  if (idx < input_length) {
+    const T x = input[idx];
+    const T in = (bias == nullptr) ? x : (x + bias[idx % bias_length]);
+    const T cdf = a + a * Tanh(in * (c * in * in + b));
+    output[idx] = in * cdf;
+  }
+}
+
+template <unsigned TPB>
+__global__ void FastGeluKernel2(const half2 a, const half2 b, const half2 c, int input_length, int bias_length, const half2* input, const half2* bias, half2* output) {
+// half2 Arithmetic functions requires cuda architecture >= 5.3
+#if __CUDA_ARCH__ >= 530
+  const int idx = blockIdx.x * TPB + threadIdx.x;
+  if (idx < input_length) {
+    const half2 x = input[idx];
+    const half2 in = (bias == nullptr) ? x : (x + bias[idx % bias_length]);
+    const half2 cdf = a + a * Tanh(in * (c * in * in + b));
+    output[idx] = in * cdf;
+  }
+#endif
+}
+
+template <>
+bool LaunchFastGeluKernel(cudaStream_t stream, int input_length, int bias_length, const float* input, const float* bias, float* output) {
+  constexpr int blockSize = 256;
+  const int gridSize = (input_length + blockSize - 1) / blockSize;
+  FastGeluKernel<float, blockSize><<<gridSize, blockSize, 0, stream>>>(A, B, C, input_length, bias_length, input, bias, output);
+
+  return CUDA_CALL(cudaPeekAtLastError());
+}
+
+template <>
+bool LaunchFastGeluKernel(cudaStream_t stream, int input_length, int bias_length, const half* input, const half* bias, half* output) {
+  constexpr int blockSize = 256;
+
+  if (0 == (bias_length & 1) && DeviceProp::GetDeviceProps().major >= 7) {
+    const int n = input_length / 2;
+    const int gridSize = (n + blockSize - 1) / blockSize;
+    const half2 A2 = __floats2half2_rn(A, A);
+    const half2 B2 = __floats2half2_rn(B, B);
+    const half2 C2 = __floats2half2_rn(C, C);
+    const half2* input2 = reinterpret_cast<const half2*>(input);
+    const half2* bias2 = reinterpret_cast<const half2*>(bias);
+    half2* output2 = reinterpret_cast<half2*>(output);
+    FastGeluKernel2<blockSize><<<gridSize, blockSize, 0, stream>>>(A2, B2, C2, n, bias_length / 2, input2, bias2, output2);
+  } else {
+    const int gridSize = (input_length + blockSize - 1) / blockSize;
+    FastGeluKernel<half, blockSize><<<gridSize, blockSize, 0, stream>>>(A, B, C, input_length, bias_length, input, bias, output);
+  }
```

Comment:
should be able to merge FastGeluKernel2 & FastGeluKernel into one method, move the code like __floats2half2_rn inside the #if __CUDA_ARCH__ >= 530.
also why half2 is required?

Reply:
It is able to merge them but that will make kernel more complex. Current choice is to make kernel as simple as possible.

Half2 kernel is faster than half kernel on v100, and it is optimization needed to get best performance on v100.